evm_arithmetization/src/prover.rs

```diff
@@ -30,132 +30,136 @@ use starky::stark::Stark;
 
 use crate::all_stark::{AllStark, Table, NUM_TABLES};
 use crate::cpu::kernel::aggregator::KERNEL;
-use crate::cpu::kernel::interpreter::{generate_segment, ExtraSegmentData};
+use crate::cpu::kernel::interpreter::{
+    generate_segment, generate_segment_given_data, ExtraSegmentData,
+};
 use crate::generation::state::GenerationState;
 use crate::generation::{generate_traces, GenerationInputs, MemBeforeValues, SegmentData};
 use crate::get_challenges::observe_public_values;
 use crate::memory::segments::Segment;
 use crate::proof::{AllProof, MemCap, PublicValues, RegistersData};
-use crate::witness::memory::MemoryAddress;
+use crate::witness::memory::{MemoryAddress, MemoryState};
 use crate::witness::state::RegistersState;
 
+/// Structure holding the data needed to initialize a segment.
+#[derive(Clone, Default, Debug)]
+pub struct GenerationSegmentData {
```

Comment:
Why not also adding `registers_after`? The overhead would be imo minimal, it would remove it from the argument list of the proving functions (making things a bit less cluttered) and it would remove the possible confusion with N+1 in `generate_all_data_segments`. I think it's worth it.

Reply:
I deleted `generate_segment_given_data` and put everything in `generate_all_data_segments`.